crates/core/src/host/scheduler.rs

```diff
@@ -249,52 +224,248 @@ impl SchedulerActor {
-                )
-                .await;
-            if !matches!(res, Err(ReducerCallError::NoSuchModule(_))) {
-                // if we didn't actually call the reducer because the module exited, leave
-                // the ScheduledReducer in the database for when the module restarts
-                let _ = db.remove(id.0.to_le_bytes());
-            }
-            match res {
-                Ok(_) => {}
-                Err(e) => log::error!("invoking scheduled reducer failed: {e:#}"),
+
+        if let Some(module_host) = self.module_host.upgrade() {
+            let db = module_host.dbic().relational_db.clone();
+            let ctx = ExecutionContext::internal(db.address());
+            let tx = db.begin_mut_tx(IsolationLevel::Serializable);
+            let identity = module_host.info().identity;
+
+            match get_schedule_row_mut(&ctx, &tx, &db, id.table_id, id.schedule_id) {
+                Ok(schedule_row) => {
+                    let schedule_row_ptr = schedule_row.pointer();
+
+                    match self.proccess_schedule(&ctx, &tx, &db, id, &schedule_row) {
+                        Ok(ScheduledReducer {
+                            reducer,
+                            bsatn_args,
+                            is_repeated,
+                        }) => {
+                            let _ = tokio::spawn(async move {
+                                let res = module_host
+                                    .call_reducer(
+                                        Some(tx),
+                                        identity,
+                                        None,
+                                        None,
+                                        None,
+                                        None,
+                                        &reducer,
+                                        ReducerArgs::Bsatn(bsatn_args.into()),
+                                    )
+                                    .await;
+
+                                // if we didn't actually call the reducer because the module exited, leave
+                                // the ScheduledReducer in the database for when the module restarts
+                                // and also not delete it if it's a repeated schedule
+                                if matches!(res, Err(ReducerCallError::NoSuchModule(_))) && !is_repeated {
+                                    //`call_reducer` above moves the tx, so its safe to create new here
+                                    let _ = db
+                                        .with_auto_commit(&ctx, |tx| {
+                                            db.delete(tx, id.table_id, [schedule_row_ptr]);
+                                            Ok::<(), anyhow::Error>(())
+                                        })
+                                        .map_err(|e| log::error!("deleting scheduled reducer failed: {e:#}"));
```

Comment:
Strangely, the types are lying -- this transaction cannot actually fail.

But if it could, we might run the scheduler again in the future, is that correct?

I think it's important to note both in the proposal and user-facing documentation that scheduled reducers have "at least once" semantics.

Reply:
But does that also means that there is a bug in our logic?

Reply:
I think it means that, like in many instances in the codebase, we formulate types not to express what is the case, but what might be the case in the future.

In this case: transactions may become fallible under MVCC, if and when that is supported, but they are, in fact, not.

If we resist debating the code structure, the possibility of this particular transaction being fallible means that we may run the scheduled reducer more than once.

I think that’s fine, but we need to be diligent in documenting the behavior.